src/main/flight/dyn_notch_filter.c

```diff
@@ -102,127 +106,171 @@ typedef struct peak_s {
-static peak_t FAST_DATA_ZERO_INIT     peaks[DYN_NOTCH_COUNT_MAX];
-static float FAST_DATA_ZERO_INIT      sdftData[SDFT_BIN_COUNT];
-static float FAST_DATA_ZERO_INIT      sdftSampleRateHz;
-static float FAST_DATA_ZERO_INIT      sdftResolutionHz;
-static int FAST_DATA_ZERO_INIT        sdftStartBin;
-static int FAST_DATA_ZERO_INIT        sdftEndBin;
-static float FAST_DATA_ZERO_INIT      sdftMeanSq;
-static float FAST_DATA_ZERO_INIT      dynNotchQ;
-static float FAST_DATA_ZERO_INIT      dynNotchMinHz;
-static float FAST_DATA_ZERO_INIT      dynNotchMaxHz;
-static uint16_t FAST_DATA_ZERO_INIT   dynNotchMaxFFT;
-static float FAST_DATA_ZERO_INIT      gain;
-static int FAST_DATA_ZERO_INIT        numSamples;
-
-void gyroDataAnalyseInit(gyroAnalyseState_t *state, const uint32_t targetLooptimeUs)
-{
-    // initialise even if FEATURE_DYNAMIC_FILTER not set, since it may be set later
-    dynNotchQ = gyroConfig()->dyn_notch_q / 100.0f;
-    dynNotchMinHz = gyroConfig()->dyn_notch_min_hz;
-    dynNotchMaxHz = MAX(2 * dynNotchMinHz, gyroConfig()->dyn_notch_max_hz);
+typedef struct state_s {
+
+    // state machine step information
+    int tick;
+    int step;
+    int axis;
+
+} state_t;
+
+typedef struct dynNotch_s {
 
-    // gyroDataAnalyse() is running at targetLoopRateHz (which is PID loop rate aka. 1e6f/gyro.targetLooptimeUs)
-    const float targetLoopRateHz = 1.0f / targetLooptimeUs * 1e6f;
-    numSamples = MAX(1, targetLoopRateHz / (2 * dynNotchMaxHz)); // 600hz, 8k looptime, 6.00
+    float q;
+    float minHz;
+    float maxHz;
+    int count;
 
-    sdftSampleRateHz = targetLoopRateHz / numSamples;
+    uint16_t maxCenterFreq;
+    float centerFreq[XYZ_AXIS_COUNT][DYN_NOTCH_COUNT_MAX];
+    
+    uint32_t looptimeUs;
+    biquadFilter_t notch[XYZ_AXIS_COUNT][DYN_NOTCH_COUNT_MAX];
+
+} dynNotch_t;
+
+// dynamic notch instance
+static FAST_DATA_ZERO_INIT dynNotch_t dynNotch;
+
+// accumulator for oversampled data => no aliasing and less noise
+static FAST_DATA_ZERO_INIT int   sampleIndex;
+static FAST_DATA_ZERO_INIT int   sampleCount;
+static FAST_DATA_ZERO_INIT float sampleCountRcp;
+static FAST_DATA_ZERO_INIT float sampleAccumulator[XYZ_AXIS_COUNT];
+
+// downsampled data for frequency analysis
+static FAST_DATA_ZERO_INIT float sampleAvg[XYZ_AXIS_COUNT];
+
+// parameters for peak detection and frequency analysis
+static FAST_DATA_ZERO_INIT state_t state;
+static FAST_DATA_ZERO_INIT sdft_t  sdft[XYZ_AXIS_COUNT];
+static FAST_DATA_ZERO_INIT peak_t  peaks[DYN_NOTCH_COUNT_MAX];
+static FAST_DATA_ZERO_INIT float   sdftData[SDFT_BIN_COUNT];
+static FAST_DATA_ZERO_INIT float   sdftSampleRateHz;
+static FAST_DATA_ZERO_INIT float   sdftResolutionHz;
+static FAST_DATA_ZERO_INIT int     sdftStartBin;
+static FAST_DATA_ZERO_INIT int     sdftEndBin;
+static FAST_DATA_ZERO_INIT float   sdftMeanSq;
+static FAST_DATA_ZERO_INIT float   gain;
+
+PG_REGISTER_WITH_RESET_FN(dynNotchConfig_t, dynNotchConfig, PG_DYN_NOTCH_CONFIG, 0);
```

Comment:
All of the parameter group related functionality should really go into their own files in `src/main/pg/`, as is the case for all other non-legacy parameter groups.